include/vinecopulib/bicop/implementation/abstract.ipp

```diff
@@ -118,25 +129,116 @@ AbstractBicop::set_loglik(const double loglik)
 {
   loglik_ = loglik;
 }
+
+inline void
+AbstractBicop::set_var_types(const std::vector<std::string>& var_types)
+{
+  if (var_types.size() != 2) {
+    throw std::runtime_error("var_types must have size two.");
+  }
+  var_types_ = var_types;
+}
 //! @}
 
 //! evaluates the pdf, but truncates it's value by DBL_MIN and DBL_MAX.
 //! @param u matrix of evaluation points.
 inline Eigen::VectorXd
-AbstractBicop::pdf(const Eigen::Matrix<double, Eigen::Dynamic, 2>& u)
+AbstractBicop::pdf(const Eigen::MatrixXd& u)
 {
+
+  Eigen::VectorXd pdf(u.rows());
+  if (var_types_ == std::vector<std::string>{ "c", "c" }) {
+    pdf = pdf_raw(u.leftCols(2));
+  } else if (var_types_ == std::vector<std::string>{ "d", "d" }) {
+    pdf = pdf_d_d(u);
+  } else {
+    pdf = pdf_c_d(u);
+  }
+
   auto trim = [](const double& x) {
     return std::min(DBL_MAX, std::max(x, DBL_MIN));
   };
-  return tools_eigen::unaryExpr_or_nan(pdf_raw(u), trim);
+  return tools_eigen::unaryExpr_or_nan(pdf, trim);
+}
+
+inline Eigen::VectorXd
+AbstractBicop::pdf_c_d(const Eigen::MatrixXd& u)
+{
+  auto umax = u.leftCols(2);
```

Comment:
Do we really have to do all this copying? Same question for all the functions below...